### PR TITLE
#5612: Add support for Selector Templates in Selector Hints

### LINKED ISSUE #5612: Slice 1.1.2: Refine selector hints for Salesforce (ancestor selector stencil for Case Details)
<https://www.notion.so/pixiebrix/Improve-Element-Selector-Inference-f2f1d961e955404eb0287831ca50c47e?pvs=4#8e24e2e2078c431aadd0e872d2588884>

<span data-preserve-white-space></span>  
**Sub Slice 1\.1\.2:** Define ancestor selector stencil for Case Details

### PR DESCRIPTION
**Pull request by turbochef**
## What does this PR do?

Closes #5612

Allows us to add template selector hints. Ex:

HTML
```html
<div class="parent">
  <span class="label-class">Identifying label</span>
  <span class="selected-item">1234</span>
</div>
```

Template
```json
{
  "template": ".container:has(.label:contains({{ label.text }}))"
  "selector": ".slds-form-element:has(.parent)", 
  "extract": {
		"label": ".label-class"
  },
}
```

Resulting selector
```css
.container:has(.label:contains(Identifying label))
```

## Discussion

- If a both a required selector and a template apply to an elements ancestor. I'm applying the template.

## Checklist

- [x] Add tests
- [x] Designate a primary reviewer

### CODE REVIEW

File: src/utils/inference/siteSelectorHints.ts [hunk 1]
```diff
@@ -60,6 +67,16 @@ export const SELECTOR_HINTS: SiteSelectorHint[] = [
       /^\[name='leftsidebar'] */,
     ],
     requiredSelectors: ['[role="main"]>.active'],
+    selectorTemplates: [
+      {
+        template:
+          ".slds-form-element:has(.test-id__field-label:contains({{ label.text }}))",
```
Comment (twschiller) [inline src/utils/inference/siteSelectorHints.ts:73]: @turbochef my template had an error here. There needs to be quote marks for this to work I think

`:contains("{{ label.text }})")`

Comment (twschiller) [reply src/utils/inference/siteSelectorHints.ts:73]: https://api.jquery.com/contains-selector/

Comment (turbochef) [reply src/utils/inference/siteSelectorHints.ts:73]: Your sketch included them I think. I'll add them. It did seem to work without them though.

File: src/utils/inference/selectorInference.ts [hunk 2]
```diff
@@ -428,19 +437,67 @@ function findAncestorsWithIdLikeSelectors(
...
+    }
+
+    const closestAncestorElement = $(element)
+      .closest(`:has(${extractRule})`)
+      // eslint-disable-next-line -- jQuery false positive
+      .find(extractRule);
+
+    return {
+      text: closestAncestorElement.text(),
+      attr: getElementAttributes(closestAncestorElement[0]),
```
Comment (twschiller) [inline src/utils/inference/selectorInference.ts:474]: The original sketch was to spread these out. I need to see how the naming works though. I might make a tweak before I merge

**Comment by twschiller** [comment]
@turbochef at first glance this looks good. I'll manually QA/test it out and then approve

File: src/utils/inference/selectorInference.ts (hunk 2 above)
Comment (turbochef) [reply src/utils/inference/selectorInference.ts:474]: I forgot to spread them since we didn't have them in a rule. I'll update it.

**Comment by twschiller** [comment]
@turbochef I made some tweaks here (see my commit)

- Moved the test hint into the test file
- Added logic to account for templates generating non-unique selectors in the case of substrings

**Review by twschiller — APPROVED** [review]

**Pull request merged**